Redesign notice handling and introduce `NoticeProcessorPipeline`

Him188 opened the pull request:
- 优化内部实现代码，去重，~dump trash~
- 分割 decode 和处理: `PacketFactory` 处理同步与解析等逻辑, 将结果提交给 `NoticeProcessorPipeline`
- 按事件类型分类处理来自服务器事件, 如 `GroupListNoticeProcessor`
- 允许写测试
- 增加可读性
- 顺便修复恢复群主身份时的事件
- 顺便修复 StrangerAddedEvent 重复广播问题
- 顺便修复 Stranger 没有 close 的问题
- 顺便修复一些 Contact 的 scope parent 问题
- 顺便修复 IMirai.newStranger 的安全性问题

Him188 committed:
Add contracts to `cast`, `safeCast` and `castOrNull`; Add `uncheckedCast`

Him188 committed:
Add `out` variance to generic type of `ContactList`

Him188 committed:
Use `NormalMemberImpl` for properties of `GroupImpl`

Him188 committed:
Use internal types in implementations of properties in `Bot`

Him188 committed:
Add `TypeSafeMap`

Him188 committed:
Redesign notice handling and introduce `NoticeProcessorPipeline`

Him188 committed:
Fix incorrect coroutine scope created for contacts in `IMirai.newFriend` and `IMirai.newStranger`

Him188 committed:
Redesign notice handling and introduce `NoticeProcessorPipeline` part 2

Do not broadcast StrangerAddedEvent if added twice, fix stranger scope not closed

Do not add new instance if there is already one

Close and remove corresponding stranger instance if there is new friend.

Him188 committed:
Add `UnconsumedNoticesAlerter` and cleanup

Co-authored-by: Karlatemp <karlatemp@vip.qq.com>

Him188 committed:
Improve performance

Karlatemp commented on mirai-core-utils/src/commonMain/kotlin/TypeSafeMap.kt:
```diff
@@ -0,0 +1,43 @@
+        map[key] = value
+    }
+
+    public fun <T> remove(key: TypeKey<T>): T? = map.remove(key).uncheckedCast()
```
```suggestion
    public fun <T> remove(key: TypeKey<T>): T? = map.remove(key)?.uncheckedCast()
```

Karlatemp commented on mirai-core/src/commonMain/kotlin/contact/GroupImpl.kt:
```diff
@@ -289,16 +292,43 @@ internal fun Group.newMember(memberInfo: MemberInfo): Member {
+        return AnonymousMemberImpl(this, this.coroutineContext, memberInfo)
+    }
+    return null
+}
```
我认为在 `memberInfo.anonymousId` 不存在的时候抛出一个错误更好

Karlatemp commented on mirai-core/src/commonMain/kotlin/utils/io/serialization/utils.kt:
```diff
@@ -158,7 +158,11 @@ internal fun <T : ProtoBuf> T.toByteArray(serializer: SerializationStrategy<T>):
+internal fun <T : ProtoBuf> ByteArray.loadAs(deserializer: DeserializationStrategy<T>, offset: Int = 0): T {
+    if (offset != 0) {
+        require(this.size >= offset) { "size < offset" }
+        return this.copyOfRange(offset, this.lastIndex).loadAs(deserializer)
```
`lastIndex` 的定义是 `size - 1`, 根据 `java.util.Arrays.copyOfRange(<ArrayType>, int from, int to)`, 最后的 `to` 应该是 `array.size`, 此处根据写 Java 的数组复制的经验这行会导致复制少了最后一字节

Karlatemp commented on mirai-core/src/commonMain/kotlin/AbstractBot.kt:
```diff
@@ -89,13 +89,18 @@ internal abstract class AbstractBot constructor(
+    final override val asStranger: StrangerImpl by lazy {
+        Mirai.newStranger(this, StrangerInfoImpl(bot.id, bot.nick)).cast()
+    }
+    final override var nick: String by asFriend.info::nick
```
`nick` 在登录完成后会进行最后一次初始化校检, 使用的是 `try-catch`, 现在不是 `lateinit var` 了会导致最后的初始化校检不执行

Him188 replied:
Where? 估计在网路重构时删了
![image](https://user-images.githubusercontent.com/12100985/123628908-3301d300-d846-11eb-9d7b-4e97730cc5c2.png)

Karlatemp replied:
还真没了
https://github.com/mamoe/mirai/issues/1023
https://github.com/mamoe/mirai/commit/f12f78a55fa8725fdd03d3687823c8aa8838fccc

sandtechnology replied:
实际上 nick 应该不存在这种问题了 传过来的 tlv 会告诉你 nick，不过还是加上比较好

Him188 committed:
Remove `Group.newAnonymousMember`

Him188 committed:
Fix `ByteArray.loadAs`

Him188 committed:
Redesign notice handling and introduce `NoticeProcessorPipeline` part 3:

- Extract `SyncController`;
- Add super type `ProtocolStruct` to `ProtoBuf` and `JceStruct` to standardize `NoticeProcessorPipeline.process`
- and other implementations for pipeline

Him188 committed:
Extract utils into multiple files

Him188 committed:
apiDump

Him188 committed:
Delete empty ProtoBuf.kt

Him188 commented:
现在是架构差不多完了，还差具体协议实现的测试和一些小细节。正等待网络层那边实现一个 recording/replaying 来方便做标准化测试（2.7-RC 进行中）

Him188 committed:
Move `bkn` from `QQAndroidBot` to `WLoginSigInfo`

Him188 committed:
Remove outdated todo in `FriendNoticeProcessor`

Him188 committed:
Add `MutableProcessResult`; remove locks

Him188 committed:
Migrate more transformers to `NoticeProcessorPipeline`

Him188 committed:
Migrate more transformers to `NoticeProcessorPipeline`

Him188 committed:
Ensure all items in `MultiPacket` are visited

Him188 committed:
Remove `.debug` in `mirai.network.debug.notice.pipeline.log.full`

Him188 committed:
Avoid boxing in `TypeSafeMap` and support serialization

Him188 committed:
Add simple constructor for `TypeSafeMap`

Him188 committed:
Add `TraceLoggingNoticeProcessor`

Him188 committed:
Make `UnconsumedNoticesAlerter` disabled by default

Him188 committed:
Fix `EventBroadcasterPacketHandler.impl`

Him188 committed:
Add `AbstractPipelineContext`

Him188 committed:
Add `RecordingNoticeProcessor`

Him188 committed:
Add gradle configuration "use.maven.local"

Him188 committed:
Fix `ParseErrorPacket` not logged

Him188 committed:
Add `toString` for TypeSafeMap implementations

Him188 committed:
Make `TypeSafeMap()` returns `TypeSafeMap.EMPTY`

Him188 committed:
Add util `readResource`

Him188 committed:
Add util `CollectionDiff`

Him188 committed:
Add util `Array<Byte>.toUHexString`

Him188 committed:
Add ConstructorCallCodegen

Him188 committed:
Redesign packet recording

Him188 committed:
Fix group transfer and retrieve, add tests for notice events:
- GroupRetrieveTest
- GroupTransferTest
- MemberAdminChangeTest
- MemberQuitTest
- MuteTest

Him188 committed:
Use yamlkt 0.10.2

Him188 committed:
Fix `MemberJoinEvent` processing and add `MemberJoinTest`

Him188 committed:
Add test for message events

Him188 committed:
Fix `RecordingNoticeProcessorTest`

Him188 committed:
Get group by uin in `OnlineMessageSourceFromTempImpl`, fix the problem that some temp messages could not be received, fix #1410

Him188 committed:
Add comments for kick event

Him188 committed:
Handle possible `MemberLeaveEvent.Kick` for `NewContact.SystemMsgNewGroup, subType=5, groupMsgType=6`, fix #1171

Him188 committed:
Ignore `NewContact.SystemMsgNewGroup, subType=0, groupMsgType=8`, fix #1388

Him188 committed:
Ignore `NewContact.SystemMsgNewGroup, subType=5, groupMsgType=16`, fix #1467

Him188 committed:
Support nested structures in desensitizer

Him188 committed:
Remove more verbose default values in ConstructorCallCodegen

Him188 committed:
Add `BotInvitedJoinTest`

Him188 committed:
Use `getGroupByUinOrCodeOrFail` in OnlineMessageSourceFromTempImpl for tests

Karlatemp commented on mirai-core/src/commonMain/kotlin/network/notice/priv/FriendNoticeProcessor.kt:
```diff
@@ -0,0 +1,290 @@
+        }
+    }
+
+    private fun NoticePipelineContext.handleAvatarChanged(body: ModCustomFace) {
+        if (body.uin == bot.id) {
+            collect(BotAvatarChangedEvent(bot))
+        }
+        collect(FriendAvatarChangedEvent(bot.getFriend(body.uin) ?: return))
+    }
```
```suggestion
    private fun NoticePipelineContext.handleAvatarChanged(body: ModCustomFace) {
        if (body.uin == bot.id) {
            collect(BotAvatarChangedEvent(bot))
        } else {
            collect(FriendAvatarChangedEvent(bot.getFriend(body.uin) ?: return))
        }
    }
```

Him188 replied:
Have you run any tests? `RunRecorderKt` maybe helpful.

Karlatemp replied:
No, but I think that logic is the correct logic

Him188 committed:
Fix FriendAvatarChangedEvent

Co-authored-by: Karlatemp <karlatemp@vip.qq.com>
